text/0000-profile-features.md

```diff
@@ -0,0 +1,145 @@
+- Feature Name: profile_features
+- Start Date: 22-03-2017
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Support profile-specific overrides for cargo features and dependencies.
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes, users wish to control the build of their crates in a manner which
+is dependent on which profile it is being built for, because of the differences
+in use cases between the profiles. Today, this is not easy to achieve directly,
+and instead requires hacking with the existing `cfg` options that roughly
+correspond to some of the profiles (such as `test` and `debug_assertions`).
+
+This system is ad hoc and doesn't allow downstream users to effectively control
+the way these features are used by their dependencies. To solve this problem,
+we allow profiles to play a first class role in feature and dependency
+resolution.
+
+# Detailed design
+[design]: #detailed-design
+
+Each profile gains these additional members:
+
+* `features`
+* `dependencies`
+* `dev-dependencies`
+* `build-dependencies
+* `target`
+
+Each of these corresponds to the same top-level object. When compiling in
+a profile, its members under these objects are merged with the top level
+equivalent, overriding any overlapping members. This enables users to configure
+thier builds differently in each profile if necessary.
+
+Specifically:
+
+* The `features` table for a profile is *merged* with the base `features`
+table; each member of the `features` table defined in the profile-specific
+table *replaces* the member in the base table (they are not merged together).
```

Comment:
Merging the table means that you don't need to repeat all of its elements. For example:

```
[features]
foo = []
bar = []

[profile.dev.features]
bar = ["foo"]
```

If we *replaced* the table, this wouln't be sensical because there would no longer be a `foo` feature.